linear_algebra/tensor_product.lean

```diff
@@ -0,0 +1,378 @@
+/-
+Copyright (c) 2018 Kenny Lau. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Kenny Lau
+
+Tensor product of modules over commutative rings.
+
+-/
+
+import group_theory.free_abelian_group
+import linear_algebra.linear_map_module
+
+universes u v w u₁ v₁
+
+variables {R : Type u} [comm_ring R]
+variables (M : Type v) (N : Type w) (P : Type u₁) (Q : Type v₁)
+variables [module R M] [module R N] [module R P] [module R Q]
+include R
+
+section bilinear
+
+variables {M N P Q}
+structure is_bilinear_map {M N P}
+  [module R M] [module R N] [module R P]
+  (f : M → N → P) : Prop :=
+(add_pair : ∀ x y z, f (x + y) z = f x z + f y z)

```

Comment:
instead of `_pair` or `pair_` use `add_right`, `add_left`, `smul_left` and smul_right`.
This should be a closer match to our standard notation.

Reply:
I still don't know which one is right and which one is left.

Reply:
It's `left` when the addition or scalar-multiplication is applied to the left argument. Or you say `left` if the argument in which the function is supposed to be linear is the left argument.

Does this clarify the rule?

Reply:
So e.g. `add_pair` -> `add_left`, `pair_add` -> `add_right`

Reply:
if `add_pair` = `add_left`, why is `add_mul` = `right_distrib`?